### PR TITLE
Corr 3: add corr for arrays

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by jbloom22** [comment]
will review once corr 2 is in

**Comment by jbloom22** [comment]
great pls rebase

**Comment by tpoterba** [comment]
rebased

File: hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala [hunk 1]
```diff
@@ -306,5 +307,73 @@ object ArrayFunctions extends RegistryFunctions {
...
+
+        EmitTriplet(
+          Code(
+            setup1,
+            setup2),
+          m1 || m2 || Code(
+            l1 := t1.loadLength(region, a1),
+            l2 := t2.loadLength(region, a2),
+            l1.cne(l2).mux(
+              Code._fatal(new CodeString("'corr': cannot compute correlation between two arrays of different length: ")
```
Comment (jbloom22) [inline hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala]: "between arrays with different lengths"

File: hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala [hunk 2]
```diff
@@ -306,5 +307,73 @@ object ArrayFunctions extends RegistryFunctions {
...
+                .concat(l1.toS)
+                .concat(", ")
+                .concat(l2.toS)),
+              l1.ceq(0))),
```
Comment (jbloom22) [inline hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala]: what effect does this line have?

File: hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala [hunk 3]
```diff
@@ -306,5 +307,73 @@ object ArrayFunctions extends RegistryFunctions {
...
+              Code(
+                (t1.isElementDefined(region, a1, i) && t2.isElementDefined(region, a2, i)).mux(
+                  Code(
+                    x := region.loadDouble(t1.loadElement(region, a1, i)),
+                    y := region.loadDouble(t2.loadElement(region, a2, i)),
+                    xSum := xSum + x,
+                    ySum := ySum + y,
+                    xSqSum := xSqSum + (x * x),
+                    ySqSum := ySqSum + (y * y),
+                    xySum := xySum + (x * y),
```
Comment (jbloom22) [inline hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala]: I figure doing the x stuff first, and then the y stuff, couldn't hurt register usage

File: hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala [hunk 4]
```diff
@@ -306,5 +307,73 @@ object ArrayFunctions extends RegistryFunctions {
...
+            xySum := 0d,
+            Code.whileLoop(i < l1,
+              Code(
+                (t1.isElementDefined(region, a1, i) && t2.isElementDefined(region, a2, i)).mux(
+                  Code(
+                    x := region.loadDouble(t1.loadElement(region, a1, i)),
+                    y := region.loadDouble(t2.loadElement(region, a2, i)),
+                    xSum := xSum + x,
+                    ySum := ySum + y,
+                    xSqSum := xSqSum + (x * x),
```
Comment (jbloom22) [inline hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala]: parens seem like overkill if you're not going to use them in the final expression below as well

File: hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala (hunk 2 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala]: returns missing if the length is 0

**Pull request merged**